imap/caldav_alarm.c

```diff
@@ -865,20 +838,169 @@ static void process_one_record(struct mailbox *mailbox, uint32_t imap_uid,
 
     syslog(LOG_DEBUG, "processing per-user alarms");
 
-    mailbox_get_annotate_state(mailbox, record.uid, NULL);
-    annotatemore_findall(mailbox->name, record.uid, PER_USER_CAL_DATA,
+    mailbox_get_annotate_state(mailbox, record->uid, NULL);
+    annotatemore_findall(mailbox->name, record->uid, PER_USER_CAL_DATA,
                          /* modseq */ 0, &process_peruser_alarms_cb,
                          &prock, /* flags */ 0);
 
     data.lastrun = runtime;
-    write_lastalarm(mailbox, &record, &data);
+    write_lastalarm(mailbox, record, &data);
 
-    update_alarmdb(mailbox->name, record.uid, data.nextcheck);
+    update_alarmdb(mailbox->name, record->uid, data.nextcheck);
 
 done_item:
     if (ical) icalcomponent_free(ical);
 }
 
+static void process_futurerelease(struct mailbox *mailbox,
+                                    struct index_record *record)
+{
+    struct buf buf = BUF_INITIALIZER;
+    json_t *submission = NULL, *identity, *envelope;
+    int r = 0;
+
+    syslog(LOG_DEBUG, "processing future delivery");
+
+    if ((record->internal_flags & FLAG_INTERNAL_EXPUNGED) ||
+        (record->system_flags & FLAG_DELETED))
+        goto done;
+
+    /* Load message */
+    message_t *m = message_new_from_record(mailbox, record);
+    const char *fname;
+    r = message_get_fname(m, &fname);
+    if (r) {
+        syslog(LOG_ERR, "process_futurerelease: can't get fname for %s:%u",
+               mailbox->name, record->uid);
+        goto done;
+    }
+
+    int msgfd = open(fname, 0);
+    if (msgfd == -1) {
+        syslog(LOG_ERR, "process_futurerelease: can't open %s: %m", fname);
+        goto done;
+    }
+
+    struct stat sbuf;
+    if (fstat(msgfd, &sbuf) == -1) {
+        syslog(LOG_ERR, "process_futurerelease: can't fstat %s: %m", fname);
+        close(msgfd);
+        goto done;
+    }
+
+    buf_init_mmap(&buf, 1, msgfd, fname, sbuf.st_size, mailbox->name);
+    close(msgfd);
+    if (!buf_len(&buf)) {
+        syslog(LOG_ERR, "process_futurerelease: can't mmap %s: %m", fname);
+        goto done;
+    }
+
+    /* Parse the submission object from the first header field */
+    json_error_t jerr;
+    size_t hdrlen = strlen(JMAP_SUBMISSION_HDR) + 1; /* +1 for ':' */
+    submission = json_loadb(buf_base(&buf) + hdrlen, buf_len(&buf) - hdrlen,
+                            JSON_DISABLE_EOF_CHECK, &jerr);
+    if (!submission) {
+        syslog(LOG_ERR,
+               "process_futurerelease: failed to parse submission obj");
+        goto done;
+    }
+    envelope = json_object_get(submission, "envelope");
+    identity = json_object_get(submission, "identityId");
+
+    /* Trim submission header field */
+    const char *p = strstr(buf_base(&buf), "}\r\n") + 3; /* +3 for '}\r\n' */
```

Comment:
This feels dangerous in conjunction with arbitrary content and presumably header wrapping - we could wind up not removing the entire header